keps/sig-scheduling/20200114-multi-scheduling-profiles.md

```diff
@@ -0,0 +1,202 @@
+
+We propose making the scheduler run different framework plugin configurations,
+which we will call profiles and will be associated to a scheduler name.
+Pods can choose to be scheduled under a particular configuration by setting the
+scheduler name associated to it in its pod spec. They will continue to be
+scheduled under the default configuration if they don't specify a scheduler
+name. The scheduler will continue to schedule one pod at a time.
+
+## Motivation
+
+Clusters run a variety of workloads, which can be *broadly* classified as
+services and batch jobs. Some users may chose to run only one class of workloads
+in a cluster, so they can provide a reasonable configuration that suits their
+scheduling needs.
+
+However, users may choose to run more heterogeneous workloads in a single
+cluster. They could have a multi-tenant cluster or they might want to take
+advantage of under-utilized nodes. Or they could have a set of fixed nodes
+and a set that auto-scales.
+
+Pods can influence scheduling decisions with features such as node/pod affinity,
+tolerations or (alpha) even pod spreading. But there are 2 problems:
+
+- kube-scheduler also calculates a set of default scores that can compete with
+  these requests.
+- Authors of the workloads need to be aware of the cluster characteristics.
+
+For this reason, some operators choose to run multiple schedulers, whether those
+are different binaries or kube-scheduler with a different configuration. But
+this setup might cause race conditions between the multiple schedulers, as they
+might have a different view of the cluster resources at a given time.
+Additionally, more binaries requires more management effort.
+
+Instead, having a single kube-scheduler run multiple profiles can solve the
+users' needs avoiding the problem of multiple schedulers.
+
+### Goals
+
+- Add support in the kube-scheduler's component config API for multiple
+scheduling profiles.
+- Make kube-scheduler schedule pods using different profiles given the scheduler
+name specified in the pod spec.
+
+### Non-Goals
+
+- Introduce new default scheduling profiles.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+I have two types of workloads that I want to run in two different sets of nodes.
+For one type of workload, I want them to spread in the topology. But for the
+other type, I prefer them to get scheduled in a few nodes as possible.
+
+### Implementation Details/Notes/Constraints
+
+#### Component Config API
+
+[`v1alpha1` component config](
+https://github.com/kubernetes/kubernetes/blob/50deec2/pkg/scheduler/apis/config/types.go#L46)
+looks like the following:
+
+```go
+type KubeSchedulerConfiguration struct {
+   ...
+   SchedulerName string
+   HardPodAffinitySymmetricWeight
+   Plugins *Plugins
+   PluginConfig []PluginConfig
+   ...
+}
+```
+
+We will introduce `v1alpha2`, with the following structure:
+
+```go
+type KubeSchedulerConfiguration struct {
+   ...
+   Profiles []KubeSchedulerProfile
```

Comment:
describe how the CLI flag bindings would interact with this structure (if I start a scheduler without a config file, and continue providing flags, does that mean we will construct a scheduler configuration containing a single profile item and bind the CLI flags to subfields of that?)

Reply:
describe additional validation (e.g. schedulerName must be valid, non-empty, non-duplicate, etc)

Reply:
I also added a "defaults" section. Thanks :)